File: src/main/java/sharev/domain/card/entity/Card.java
package sharev.domain.card.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.util.Map;
import java.util.Objects;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import sharev.base_entity.BaseTimeEntity;
import sharev.domain.account.entity.Account;
import sharev.domain.card.exception.InvalidIntroduceTemplateException;
import sharev.domain.gathering.entity.Gathering;
import sharev.domain.gathering.entity.IntroduceTemplate;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Table(name = "cards")
public class Card extends BaseTimeEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "card_id")
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "gathering_id")
    private Gathering gathering;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id")
    private Account account;

    @Column
    private Integer pinNumber;

    @Column
    private Integer templateVersion;

    @Column
    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, String> introductionText;

    public Card(Gathering gathering, Account account, Integer pinNumber) {
        this.gathering = gathering;
        this.account = account;
        this.pinNumber = pinNumber;
    }

    public void updateIntroductionText(IntroduceTemplate introduceTemplate, Integer templateVersion,
                                       Map<String, String> introductionText) {

        if (!introduceTemplate.validateIntroduce(templateVersion, introductionText)) {
            throw new InvalidIntroduceTemplateException();
        }

        this.templateVersion = templateVersion;
        this.introductionText = introductionText;
    }

    public boolean isCompleted() {
        return Objects.nonNull(this.pinNumber) && Objects.nonNull(this.introductionText);
    }
}

File: src/main/kotlin/sharev/card/domain/model/Card.kt
package sharev.card.domain.model

import sharev.card.domain.exception.CardException
import sharev.card.domain.exception.CardExceptionCode
import java.util.*

data class Card(
    val id: Long,
    val gatheringId: UUID,
    val accountId: Long,
    val accountName: String,
    val accountEmail: String,
    val pinNumber: Int?,
    val templateVersion: Int?,
    val introductionText: Map<String, String>?,
) {
    fun validateIntroductionText(
        currentTemplateVersion: Int,
        templateFields: Set<String>,
        templateVersion: Int,
        introductionText: Map<String, String>,
    ) {
        if (currentTemplateVersion != templateVersion) {
            throw CardException(CardExceptionCode.INVALID_INTRODUCE_TEMPLATE)
        }

        val introduceFields = introductionText.keys

        if (templateFields.size != introduceFields.size || introduceFields.subtract(templateFields).isNotEmpty()) {
            throw CardException(CardExceptionCode.INVALID_INTRODUCE_TEMPLATE)
        }
    }

    fun isCompleted(): Boolean = pinNumber != null && introductionText != null
}
